code/drasil-lang/lib/Language/Drasil/Chunk/Eq.hs

```diff
@@ -61,6 +64,14 @@ instance Express e => Express (QDefinition e) where
         is -> defines $ apply q (map C is)
 instance ConceptDomain (QDefinition e) where cdom = cdom . view qdQua
 
+instance TypeChecks (QDefinition Expr) Expr Space where
```

Comment:
Shouldn't there be more "type checking" going on here?

Reply:
It depends on where we want to do the type checking.

For now, I've added it as a "step" in the generator, so that it's done post-creation of QDs, IMs, etc, but before code is generated from them. This allows us to type check everything at once, and not halt Drasil on sight of the a type error (there are other ways to recover from a program panic, but they're just not as easy) so that we can find and show all at once.

I think it would be great to have it type checked before code generation too. If we have the upgraded chunkdb, then we can quickly switch this to a typed parameter system. I'm not sure if it's possible, but we might also be able to move it to LiquidHaskell conditions on the smart constructors so that it shows in our code editors and compile time?

So, now what's going on here? "TypeChecks" is a typeclass that is used to expose all things (here, Expr) that need to be type-checked against a type (here, Space). We use it to collect all expressions at once, and then perform type-checking just before code generation. If/when we move to type checking on the individual chunk creation level, then we might remove this typeclass entirely.

Reply:
I think what threw me off was the name of this class. You do have more 'traditional' type checking/infer stuff in other places. So it might just be a naming thing wrt what this class actually expresses.

Reply:
I can certainly see how it has a misleading name. Do you have suggestions for it? Perhaps 'NeedsTypeChecking'?

Reply:
So this typeclass expresses a relation `Foo a b c` where it looks like we treat `a` as a container of `b`s whose types should come from `c`(and need checking). `RequiresChecking` ? I think that's slightly better than `NeedsTypeChecking`.

Reply:
Sounds better! I changed it now. Thank you.